### PR TITLE
Contract penalty

### PR DESCRIPTION
**Pull request by GalaxySciTech**
- Add hardhat framework support
- Add patch
- Add contract deployment script
- Modify HeaderReader to library
- Change the solidity version to 0.8.19
- Add test script

### CODE REVIEW

File: mainnet_contract/README.md
Comment (liam-lai) [inline mainnet_contract/README.md]: space at the front

Comment (GalaxySciTech) [reply mainnet_contract/README.md]: Got it ,The build directory was replaced with artifacts and ignored in .gitignore, I can modify it if needed

Comment (GalaxySciTech) [reply mainnet_contract/README.md]: Or do you mean to put a little human introduction on the front?

Comment (liam-lai) [reply mainnet_contract/README.md]: not... I mean you put a space at the front of yarn, ` yarn`, but it's not necessary right?

Comment (GalaxySciTech) [reply mainnet_contract/README.md]: Ur right , i will delete space👍

**Review by liam-lai — APPROVED** [review]

**Comment by liam-lai** [comment]
Good job, remember use Squash merge when you are ready to merge

**Comment by GalaxySciTech** [comment]
@liam-lai There is another thing worth noting, wanwiset said that the hexRLP return value of the node will be updated on Friday, this will affect my deployment.js, I want to wait for it to be updated before I merge

**Comment by GalaxySciTech** [comment]
> Good job, remember use Squash merge when you are ready to merge



> Good job, remember use Squash merge when you are ready to merge

Yes i do

**Comment by liam-lai** [comment]
> @liam-lai There is another thing worth noting, wanwiset said that the hexRLP return value of the node will be updated on Friday, this will affect my deployment.js, I want to wait for it to be updated before I merge

yes, I notice that from deployment.js. It's up to you to manage this change whether include or exclude. From my perspective, this PR is a refactor work on the deployment and framework. That hexRLP is a feature change, I will make it separated PR.

**Comment by GalaxySciTech** [comment]
> > @liam-lai There is another thing worth noting, wanwiset said that the hexRLP return value of the node will be updated on Friday, this will affect my deployment.js, I want to wait for it to be updated before I merge
> 
> yes, I notice that from deployment.js. It's up to you to manage this change whether include or exclude. From my perspective, this PR is a refactor work on the deployment and framework. That hexRLP is a feature change, I will make it separated PR.

Got it If he modifies the hexRLP at that time, I will submit another PR to modify the hexRLP. Am I right?

**Comment by liam-lai** [comment]
> > > @liam-lai There is another thing worth noting, wanwiset said that the hexRLP return value of the node will be updated on Friday, this will affect my deployment.js, I want to wait for it to be updated before I merge
> > 
> > 
> > yes, I notice that from deployment.js. It's up to you to manage this change whether include or exclude. From my perspective, this PR is a refactor work on the deployment and framework. That hexRLP is a feature change, I will make it separated PR.
> 
> Got it If he modifies the hexRLP at that time, I will submit another PR to modify the hexRLP. Am I right?

Yes, that's my point

**Pull request merged**
